### PR TITLE
[runtime] Use mono's runtime default to load config files when it's not embedded

### PR DESCRIPTION
**Pull request by Therzok**

### CODE REVIEW

**Comment by Therzok** [comment]
Not sure if this is the right fix, but it seemed to fix System.Configuration being used inside system-mono xammac apps.

**Comment by Therzok** [comment]
@VincentDondain no bug report opened, this is something that exhibited while trying to link VS4Mac with system mono xammac.

I think a simple repro would be:
XamMac app with system mono (not embedded)
Query `System.Diagnostics.Debug.Listeners` -> crash

That configuration file is needed by `System.Diagnostics.DiagnosticsConfiguration` so it loads properly.

cc @chamons

**Comment by Therzok** [comment]
If this is the right fix, I'll amend the commit to add that to the commit message.

**Comment by VincentDondain** [comment]
@Therzok you can do it when squashing

**Comment by Therzok** [comment]
Failure looks unrelated

**Comment by Therzok** [comment]
Stacktrace I had, for reference - https://gist.github.com/Therzok/d0d95a847108f508fb43bd435b5c0249

File: runtime/launcher.m [hunk 1]
```diff
@@ -497,6 +497,7 @@
 		
 #ifdef DYNAMIC_MONO_RUNTIME
 		config_path = [[NSString stringWithUTF8String: mono_runtime_prefix] stringByAppendingPathComponent: @"/etc/mono/config"];
+		machine_config_path = [[NSString stringWithUTF8String: mono_runtime_prefix] stringByAppendingPathComponent: @"etc/mono/4.5/machine.config"];
```
Comment (spouliot) [inline runtime/launcher.m]: hmm... we'll need to dig the git history behind this. E.g. there are null checks `machine_config_path` later (e.g. https://github.com/xamarin/xamarin-macios/pull/3290/files#diff-af570d5ca3d85213e06ec471ce874414L511) which seems to indicate this behaviour was intended.

The earlier comment (line 491) also seems to skip it - without mentioning why.

Comment (Therzok) [reply runtime/launcher.m]: Yes, I noticed that too, but without a `machine.config` being loaded, System.Configuration just crashes with: https://gist.github.com/Therzok/d0d95a847108f508fb43bd435b5c0249

Comment (Therzok) [reply runtime/launcher.m]: If there is a way to figure out which .net framework to use (2.0, 4.0, 4.5), we can use that instead.

Comment (rolfbjarne) [reply runtime/launcher.m]: Once upon a time the two code paths (statically linked mono vs dynamically loading system mono) were entirely different. Then the code was merged, and ifs added to separate different logic, and that's how the null check for `machine_config_path` came into existence: the code that dynamically loaded the system mono just didn't load the machine.config manually (https://github.com/xamarin/maccore/commit/cedbd33a14480403646ea58847893c2cf6aab40b)

However, I wonder if normal XM apps just don't load a `machine.config` or if they load a different one.

In any case I think this is something that can cause subtle changes in behavior, and somehow needs to be opt-in/opt-out.

Comment (Therzok) [reply runtime/launcher.m]: It seems that mmp copies over this file into the bundle:
https://github.com/xamarin/xamarin-macios/blob/af03020eef61ee4c4dd0b11a3f43abfab14a94bf/tools/mmp/driver.cs#L1770-L1780

File: runtime/launcher.m [hunk 2]
```diff
@@ -495,14 +497,9 @@
 		NSString *machine_config_path = nil;
 		NSError *error = nil;
 		
-#ifdef DYNAMIC_MONO_RUNTIME
-		config_path = [[NSString stringWithUTF8String: mono_runtime_prefix] stringByAppendingPathComponent: @"/etc/mono/config"];
-#else
+#ifndef DYNAMIC_MONO_RUNTIME
```
Comment (rolfbjarne) [inline runtime/launcher.m]: This looks good, but can you also update the comment just above, and the one here: https://github.com/xamarin/xamarin-macios/blob/bcd2cc2d94b624de83e91da08f3bc26e66224776/runtime/xamarin/launch.h#L75, so something like this:

```csharp
// 7a) [If not embedding] Don't parse any config files, leave it to the Mono runtime to load whatever it needs.
```

Comment (Therzok) [reply runtime/launcher.m]: Modified comment to `Don't parse any config files, leave it to the Mono runtime to load the defaults.`, and modified it in xamarin/launch.h too

**Comment by rolfbjarne** [comment]
build

**Comment by rolfbjarne** [comment]
Test failure are unrelated

* apitest/32-bit: https://github.com/xamarin/xamarin-macios/pull/3300#pullrequestreview-91532640
* XM/introspection/Classic: https://github.com/xamarin/maccore/issues/622
* XM/introspection/Unified: https://github.com/xamarin/xamarin-macios/issues/3303
*  mmptest: https://github.com/xamarin/maccore/issues/615

**Comment by rolfbjarne** [comment]
@spouliot @chamons any other comments?

**Pull request merged**